exercises/practice/dnd-character/tests/Tests.elm

```diff
@@ -0,0 +1,223 @@
+                test "ability modifier for score 10 is 0" <|
+                    \() ->
+                        DndCharacter.modifier 10
+                            |> Expect.equal 0
+            , skip <|
+                test "ability modifier for score 11 is 0" <|
+                    \() ->
+                        DndCharacter.modifier 11
+                            |> Expect.equal 0
+            , skip <|
+                test "ability modifier for score 12 is +1" <|
+                    \() ->
+                        DndCharacter.modifier 12
+                            |> Expect.equal 1
+            , skip <|
+                test "ability modifier for score 13 is +1" <|
+                    \() ->
+                        DndCharacter.modifier 13
+                            |> Expect.equal 1
+            , skip <|
+                test "ability modifier for score 14 is +2" <|
+                    \() ->
+                        DndCharacter.modifier 14
+                            |> Expect.equal 2
+            , skip <|
+                test "ability modifier for score 15 is +2" <|
+                    \() ->
+                        DndCharacter.modifier 15
+                            |> Expect.equal 2
+            , skip <|
+                test "ability modifier for score 16 is +3" <|
+                    \() ->
+                        DndCharacter.modifier 16
+                            |> Expect.equal 3
+            , skip <|
+                test "ability modifier for score 17 is +3" <|
+                    \() ->
+                        DndCharacter.modifier 17
+                            |> Expect.equal 3
+            , skip <|
+                test "ability modifier for score 18 is +4" <|
+                    \() ->
+                        DndCharacter.modifier 18
+                            |> Expect.equal 4
+            ]
+        , describe "ability"
+            [ skip <|
+                fuzzWith { runs = 1000, distribution = Test.noDistribution }
+                    (Fuzz.fromGenerator DndCharacter.ability)
+                    "generated ability should be at least 3"
+                <|
+                    \ability -> Expect.atLeast 3 ability
+            , skip <|
+                fuzzWith { runs = 1000, distribution = Test.noDistribution }
+                    (Fuzz.fromGenerator DndCharacter.ability)
+                    "generated ability should be at most 18"
+                <|
+                    \ability -> Expect.atMost 18 ability
+            , skip <|
+                fuzzWith
+                    { runs = 10000
+                    , distribution =
+                        Test.expectDistribution
+                            [ ( Test.Distribution.moreThanZero, "is 3", (==) 3 ) ]
+                    }
+                    (Fuzz.fromGenerator DndCharacter.ability)
+                    "generated ability should sometimes be 3"
+                <|
+                    \_ -> Expect.pass
```

Comment:
This is a bit more tricky. In this case we are not testing a property of the ability (the test is `Expect.pass`) but instead we are looking at the distribution. We want to make sure that the generator _sometimes_ produces a 3. `( Test.Distribution.moreThanZero, "is 3", (==) 3 )` means that we want one ability to be 3 at least once. If it's never 3, the assumption we had on the distribution is false and the test will fail.

However, getting a 3 is quite rare (0.1%), so we need to take a very large sample to make sure that we get one at least 99.99% of the time. So I crunched the numbers.

Given a specific score that appears with probability $p$, the probability that it doesn't appear is $1 - p$, the probability that it doesn't appear $n$ times in a row is $(1 - p)^n$, so the probability that it does appear at least once in $n$ runs is:
```math
1 - (1 - p)^n
```

<img width="436" alt="Screenshot 2023-10-07 at 21 54 44" src="https://github.com/exercism/elm/assets/16929078/df0ba880-ed6e-4813-a5a3-711726c56a2d">

For 3, we need 10k runs to be safe, for 18 we need 1k.
In practice, I tested getting a 3 with 1k runs and I never got it to fail, so maybe I'm missing something, possibly related to how `Test.expectDistribution` works, the docs say:
> Uses a statistical test to make sure the distribution doesn't pass or fail the distribution by accident (a flaky test). Will run more tests than specified with the runs config option if needed.

An error for a test of this type looks like
```
    Distribution report:
    ====================
      <uncategorized>:  100%  (10000x)  ██████████████████████████████
      is 3:               0%      (0x)  ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
    
    Distribution of label "is 3" was insufficient:
      expected:  more than 0%
      got:       0.000%.
    
    (Generated 10000 values.)
```

Reply:
> expectDistribution is for validating that the fuzzer passes the values you want in to the function to be tested, right?

Yes, that's exactly it.

> So I'm not sure this test is doing the right thing. Can you get it to fail by making `DndCharacter.ability` invalid.

Yes, all the example errors I showed are real, and obtained by tweaking the generators. In this case, all you need to change is to make the dice roll from 2 to 6 instead of from 1 to 6 and 3 can never be obtained.

> Maybe it is possible to use Test.Distribution to enforce the score of 3 as well, although I haven't tried it, and it looks tricky, as its a bit magic, and you need to have 4 random inputs in a row all be small. It would be good if it was possible though, and then we could use fail or onFail again to provide a good message.

I think what you are saying is that you would like a way to discard all the generated values that you are not interested in, and then run only the ones you want through the expectation. Is that right? As far as I know, it's not possible.
Or are you saying that you want to manipulate the generator to get particular values? That definitely sounds impossible.

In any case, I think the error message isn't too bad here. The test name is "generated ability should sometimes be 3" and the error message says " Distribution of label "is 3" was insufficient:".

> If we can get these tests perfect, it would also be a good learning resource for people to reference / copy paste when they do their own Random things, so we could also mention that in the instructions (if we can get them to a place where this makes sense).

I don't think a practice exercise is the right place for that. I have been thinking about a `test` concept where we wold explain how tests are structured in Elm. But testing that would be hard :)

Reply:
There is definitely something I don't understand about this, but maybe its best not to worry. It sounds like it is doing the right thing.

Maybe the test messages are ok too. They are complex, but that is something a student would have to get used to when working with random, so probably good to expose it.

Reply:
It might be good to have a testing concept, where we ask the student to write some tests, and then this could rely on that concept, and then students should be able to read the messages.
Perhaps we could update the test runner to also extract and run tests from the students code to make it happen?